lib/spack/spack/spec.py

```diff
@@ -4395,13 +4443,75 @@ def _spec_from_dict(data):
         if 'dependencies' not in node[name]:
             continue
 
-        yaml_deps = node[name]['dependencies']
-        for dname, dhash, dtypes in Spec.read_yaml_dep_specs(yaml_deps):
+        for dname, dhash, dtypes, _ in Spec.dependencies_from_node_dict(node):
             deps[name]._add_dependency(deps[dname], dtypes)
 
     return spec
 
 
+# Note: This function has been refactored from being a static method
+# of Spec to be a function at the module level. This was needed to
+# support its use in __reduce__ to pickle a Spec object in Python 2.
+# It can be moved back safely after we drop support for Python 2.7
+def _spec_from_dict(data):
+    """Construct a spec from YAML.
+
+    Parameters:
+    data -- a nested dict/list data structure read from YAML or JSON.
+    """
+    if isinstance(data['spec'], list):  # Legacy specfile format
+        return _spec_from_old_dict(data)
+
+    # Current specfile format
+    nodes = data['spec']['nodes']
+    hash_type = None
+    any_deps = False
+
+    # Pass 0: Determine hash type
+    for node in nodes:
+        if 'dependencies' in node.keys():
+            any_deps = True
+            for _, _, _, dhash_type in Spec.dependencies_from_node_dict(node):
+                if dhash_type:
+                    hash_type = dhash_type
+                    break
+
+    if not any_deps:  # If we never see a dependency...
+        hash_type = ht.dag_hash.name  # use the full_hash provenance
+    elif not hash_type:  # Seen a dependency, still don't know hash_type
+        raise spack.error.SpecError("Spec dictionary contains malformed "
+                                    "dependencies. Old format?")
+
+    hash_dict = {}
+    root_spec_hash = None
+
+    # Pass 1: Create a single lookup dictionary by hash
+    for i, node in enumerate(nodes):
+        if 'build_spec' in node.keys():
+            node_hash = node[hash_type]
+        else:
+            node_hash = node[hash_type]
```

Comment:
@alalazo Here's where the block we were discussing in #28504 was added, maybe @nhanford recalls what the intent was.  But likely not given the size of this PR.

Reply:
IIRC this is vestigial from when I was experimenting with forcing a certain hash type for spliced specs, and then I must have quickly done a "replace all" to fix it. Please feel free to remove the unnecessary conditional or let me know and I can do it.

Reply:
Thanks @nhanford I have removed the conditional in the linked PR above, we just wanted to make sure we weren't covering up the trail for someone diagnosing some potential future issue.